### PR TITLE
[stable23] Render data URL images as-is

### PR DESCRIPTION
**Pull request by DerpgonCz**
### Summary
When the image URL is a data URL, render the image as-is with `src="dataUrl"`. No need to fetch anything.

Also prevents a race condition while loading images.

Resolves: #1827

### CODE REVIEW

**Comment by DerpgonCz** [comment]
Well, what to do with this? Wait until the NC main repo updates to not be `RC`?

File: src/nodes/ImageView.vue [hunk 1]
```diff
@@ -227,6 +229,7 @@ export default {
 			this.imageLoaded = false
 			this.loaded = true
 		}
+		img.src = this.imageUrl
```
Comment (eneiluj) [inline src/nodes/ImageView.vue:232]: Why did you move this?
Is it to make sure `onload` is defined before the image is actually loaded?
I agree it's better but it works fine without moving it.
@mejo- @juliushaertl @max-nextcloud What do you think about that?

**Review by eneiluj — APPROVED** [review]
:+1: Well done, thanks a lot.
Could you make another PR for the master branch? Those changes apply perfectly to master as well.

@juliushaertl What do you think about potential XSS with SVG? Should we simply not render SVG data?

**Comment by eneiluj** [comment]
@juliushaertl Ignore my comment about XSS. It's actually safe to load base64 content in the src attribute.
Sorry for the noise.

**Comment by eneiluj** [comment]
> Well, what to do with this? Wait until the NC main repo updates to not be `RC`?

Exactly, we'll merge the approved PRs after the minor releases have been produced.

File: src/nodes/ImageView.vue (hunk 1 above)
Comment (DerpgonCz) [reply src/nodes/ImageView.vue:232]: This is due to a race condition, detailed in the issue #1827 

tldr: Assigning Image.src before Image.onload does not fire the event if the actual `load` event on the Image object happens before assigning the event handler function. (so between `onload=` and `src=`).

I understand that this is extremely unlikely to happen, maybe even impossible or really a tiny edge-case. However, better safe than pull out our hair when something like that happens.

**Comment by DerpgonCz** [comment]
> +1 Well done, thanks a lot. Could you make another PR for the master branch? Those changes apply perfectly to master as well.
> 
> @juliushaertl What do you think about potential XSS with SVG? Should we simply not render SVG data?

We could possibly check for the mimetype. Also described in the issue #1827.

File: src/nodes/ImageView.vue (hunk 1 above)
Comment (juliushaertl) [reply src/nodes/ImageView.vue:232]: Makes sense 👍

**Comment by juliushaertl** [comment]
>  Ignore my comment about XSS. It's actually safe to load base64 content in the src attribute.

While a quick test at least seems to not execute any inline JS from base64 encoded svgs, I'd rather be careful about this. Any reference mentioning that this is not an issue? In general I'd say we should avoid rendering user provided SVG especially if they might be presented to other users. 

Example payload I used:
```
<img src="data:image/svg+xml;base64,CjxzdmcgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB3aWR0aD0iMTAwIiBoZWlnaHQ9IjEwMCI+CiAgPGNpcmNsZSByPSIxMCIgY3g9IjEwIiBjeT0iMTAiIGZpbGw9ImdyZWVuIi8+CiAgPGltYWdlIGhyZWY9IngiIG9uZXJyb3I9ImphdmFzY3JpcHQ6YWxlcnQoJ1hTUycpIiAvPgo8L3N2Zz4=" />
```

**Comment by juliushaertl** [comment]
I couldn't find any good resource how different browsers process files in img tags, but if they are not parsed in terms of HTML/JS rendering it should be safe.

**Comment by DerpgonCz** [comment]
If possible, it would be a major vulnerability. Most conversations online point to it not being possible while in an `img` tag. However, opening the source directly in a browser (opening DevTools and clicking the `src` attribute and `Open link in a new tab` actually executes the code.

I am not sure how big of a vulnerability this is.

**Comment by juliushaertl** [comment]
Sounds reasonable then 👍  Also according to https://developer.mozilla.org/en-US/docs/Web/SVG/SVG_as_an_Image#restrictions this should be restricted by the browsers.

**Comment by max-nextcloud** [comment]
The freeze is over. The CI is green again.. time to get this in alongside the other pending stable23 PRs. I'm going through them one by one and hope i get to it today.
